### ISSUE TITLE
False-positive result while checking for slip zip if target path is not normalized

### ISSUE THREAD

**Issue by repolevedavaj** [srikanth-lingala/zip4j#62]
If i use non normalized target paths while unzipping, the check for slip zip ends with a false-positive result.
The issue is, that the path of the extracted file is compared with a non normalized target path in `AbstractExtractFileTask#extractFile(...)`:

` if (!new File(completePath).getCanonicalPath().startsWith(new File(outPath).getPath())) {`

This line should be changed to

`!new File(completePath).getCanonicalPath().startsWith(new File(outPath).getCanonicalPath())`

Test case (JUnit5 + AssertJ) for verification:

```
@Test
public void testUnzipFileZipSlipWithNotNormalizedTarget(@TempDir File tempDir) throws Exception {
	final File goodFile = new File(tempDir, "good.txt");
	assertThat(goodFile.createNewFile()).isTrue();
	FileUtils.write(goodFile, "good", StandardCharsets.UTF_8);

	final ZipParameters zipParameters = new ZipParameters();
	zipParameters.setFileNameInZip("good.txt");

	final ZipFile zip = new ZipFile(new File(tempDir, "test.zip"));
	zip.addFile(goodFile, zipParameters);

	zip.extractAll(new File(tempDir, "../" + tempDir.getName() + "/unzipped").getAbsolutePath());
}
```

**Comment by LeeYoung624** [comment]
I tested this case and I think it's right. Only to modify the testcase a little bit:
```
@Test
  public void testUnzipFileZipSlipWithNotNormalizedTarget() throws IOException {
    ZipFile zipFile = new ZipFile(generatedZipFile, PASSWORD);
    zipFile.addFiles(FILES_TO_ADD);

    zipFile.extractAll(new File(outputFolder.getPath(),
        ".." + InternalZipConstants.FILE_SEPARATOR + outputFolder.getName()).getAbsolutePath());

    ZipFileVerifier.verifyFolderContentsSameAsSourceFiles(outputFolder);
    verifyNumberOfFilesInOutputFolder(outputFolder, 3);
  }
```

Further more, I found other exceptions like this with test case:
```
@Test
  public void testAddFolderAndFileWithCanonicalPath() throws IOException {
    ZipFile zipFile = new ZipFile(generatedZipFile);
    ZipParameters parameters = new ZipParameters();

    String folderToAddPath = TestUtils.getTestFileFromResources("").getPath() + InternalZipConstants.FILE_SEPARATOR + ".." + InternalZipConstants.FILE_SEPARATOR + TestUtils.getTestFileFromResources("").getName();
    File folderToAdd = new File(folderToAddPath);
    zipFile.addFolder(folderToAdd, parameters);
    zipFile.addFile(TestUtils.getTestFileFromResources("file_PDF_1MB.pdf"), parameters);
    ZipFileVerifier.verifyZipFileByExtractingAllFiles(generatedZipFile, outputFolder, 13);
  }
```

I'm trying to fix all these.

**Comment by srikanth-lingala** [comment]
@LeeYoung624 Along with this change, can you please also revert all the `getCanonicalPath()` changes in [this][1] commit and revert it back to `getPath()` as it was earlier. For example on [this][2] line? Thanks.

[1]: https://github.com/srikanth-lingala/zip4j/commit/d3f304f756156e050771b1f1c859ec9e9c7cd027
[2]: https://github.com/srikanth-lingala/zip4j/commit/d3f304f756156e050771b1f1c859ec9e9c7cd027#diff-0f639ad6b7731640de24d99e7f3d05a5R179

**Comment by srikanth-lingala** [comment]
@LeeYoung624 Just to be clear, please only revert the changes in that commit and only from the tests in that commit. Thanks.

**Comment by LeeYoung624** [comment]
Sure @srikanth-lingala

**Comment by cbgxhub** [comment]
Hi srikanth-lingala 

I ran into the same Problem trying to unzip a zip file that was ziped on Windows but trying to unzip it on Linux, and came to the same conclusion regarding a fix as LeeYoung624:

 !new File(completePath).getCanonicalPath().startsWith(new File(outPath).getCanonicalPath())

Regards

**Comment by srikanth-lingala** [comment]
@cbgxhub [A pull request][1] for this issue is already open and work-in-progress. I will make a new release as soon as it is merged.

[1]: https://github.com/srikanth-lingala/zip4j/pull/63

**Comment by srikanth-lingala** [comment]
Issue fixed in v2.1.4 released today.
